onnxscript/tests/function_libs/torch_lib/ops_test.py

```diff
@@ -153,79 +154,77 @@ def run_test_output_match(
+
+                reference_torch_outputs, _ = pytree.tree_flatten(torch_output)
+                if op.name.startswith("split"):
+                    # Hack for handling split
+                    # Split returns a Sequence that should be treats as a single
+                    # value. So we wrap it into a tuple.
+                    # TODO(justinchuby): Find a more general solution
+                    reference_torch_outputs = [reference_torch_outputs]
+
+                function_output = function_executor(reference_torch_outputs)(
+                    onnx_function, input_onnx, kwargs_onnx
                 )
-
-                if op.name in ops_test_data.NONDETERMINISTIC_OPS:
-                    # Check shape and dtype only for ops that are known to be
-                    # nondeterministic
-                    test_suite.assertEqual(actual.shape, expected.shape)
-                    test_suite.assertEqual(actual.dtype, expected.dtype)
-                    continue
-
-                # Use torch.testing as opposed to np.testing to ensure dtypes and shapes match
-                try:
-                    torch.testing.assert_close(
-                        actual,
-                        expected,
-                        rtol=rtol,
-                        atol=atol,
-                        check_device=False,
+                # Finally we re-flatten everything
+                # TODO: add pytree structure comparison.
+                flattened_torch_outputs, _ = pytree.tree_flatten(torch_output)
+                flattened_function_outputs, _ = pytree.tree_flatten(function_output)
+
+                assert flattened_torch_outputs
+                assert len(flattened_torch_outputs) == len(flattened_function_outputs)
+
+                for j, (torch_output, function_output) in enumerate(
+                    zip(flattened_torch_outputs, flattened_function_outputs)
+                ):
+                    if dtype == torch.float32:
+                        # Relax atol and rtol for float32 based on empirical results
+                        # The current most relaxed values are for aten::matmul
+                        rtol = 3.7e-5
+                        atol = 1.8e-4
+                    else:
+                        rtol = None
+                        atol = None
+
+                    if not isinstance(function_output, np.ndarray):
+                        # An onnxscript tensor
+                        function_output = function_output.value
+
+                    actual = torch.tensor(function_output)
+                    expected = (
+                        torch_output
+                        if isinstance(torch_output, torch.Tensor)
+                        else torch.tensor(torch_output)
                     )
-                except AssertionError as e:
-                    if len(flattened_torch_outputs) > 1:
-                        raise AssertionError(f"Output {j} mismatch") from e
-                    raise
```

Comment:
Interested in how this raise serves this block here. It doesn't have any info though.

Reply:
https://docs.python.org/3.11/tutorial/errors.html#raising-exceptions